### ISSUE TITLE
Suggests impossible alternative

### ISSUE THREAD

**Issue by jD91mZM2** [rust-lang/rust-clippy#1772]
So, I have this:

```Rust
for i in 0..guilds.len() {
	if guilds[i].id == guild.id {
		guilds.remove(i);
		break;
	}
}
```

Clippy wants me to use
`for (i, guild) in guilds.iter().enumerate() {`

However, that won't work since `.iter().enumerate()` borrows `guilds`, so `guilds.remove` says it can't be borrowed as both immutable and mutable, or if both are mutable, it complains about that instead.

Now, I get it; it says "consider". It doesn't say I *have* to change.  
But then again, I don't wanna supress it for the whole project. Only for that file.  
Any way to do this? Any better way to do the for loop than I am currently? Any way to make clippy detect stuff like this somehow?

**Comment by killercup** [comment]
This is a valid false-positive. I'm pretty sure the lint could check whether `guilds` is being used in the loop body.

For now, you can put `#[allow(whatever_lint)]` above your function.

**Comment by legolord208** [comment]
That's what I'm doing, but it gives me a warning when I'm not using clippy then :(

**Comment by killercup** [comment]
Oh, right. You are looking for `#[cfg_attr(feature = "cargo-clippy", allow(lint_name))]` if you are using `cargo clippy`.

**Comment by legolord208** [comment]
Oh! Thanks!

**Comment by camsteffen** [comment]
There is usually a more idiomatic solution if you are iterating indices and borrowing the object in the loop. For this example, you could do this:
```rust
let pos = guilds.iter().position(|g| g.id == guild.id);
if let Some(pos) = pos {
    guilds.remove(pos);
}
```
Even though we're not using `enumerate` here, I think that is still the best general tip we can expect clippy to provide.

**Comment by legolord208** [comment]
Huh, didn't know about that

**Comment by y21** [comment]
This seems to be fixed(?).
Clippy no longer seems to complain about this AFAICT: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b13f501e95dea581a7b9a94bfb8fa547
What was the name of the lint that triggered here? I don't see it mentioned anywhere.
